File: lua/sieve_atkin.lua
N = tonumber(arg[1]) or 1000
nsqrt = N^(0.5)

is_prime = {}

--Initialize array
for i=1, N do
  is_prime[i] = false
end

for x = 1, nsqrt do
  for y = 1, nsqrt do
    n = 4 * (x^2) + y^2
    if n <= N and (n % 12 == 1 or n % 12 == 5) then
      if is_prime[n] == true then
        is_prime[n] = false
      else
        is_prime[n] = true
      end
    end
    n = 3 * (x^2) + y^2
    if n <= N and n % 12 == 7 then
      if is_prime[n] == true then
        is_prime[n] = false
      else
        is_prime[n] = true
      end
    end
    n = 3 * (x^2) - y^2
    if x > y and n <= N and n % 12 == 11 then
      if is_prime[n] == true then
        is_prime[n] = false
      else
        is_prime[n] = true
      end
    end
  end
end

for n = 5, nsqrt do
  if is_prime[n] then
    for y = n^2, N, n^2 do
      is_prime[y] = false
    end
  end
end

is_prime[2] = true
is_prime[3] = true
primes = {}

for x = 1, #is_prime - 1 do
  if is_prime[x] == true then
    table.insert(primes, x)
  end
end

for x = 1, #primes do
    print(primes[x])
end
os.exit(0)

File: python/sieve_atkin.py
import sys
import math

args = sys.argv

N = len(args) > 1 and int(sys.argv[1]) or 1000
nsqrt = int((math.pow(N, 0.5)))
is_prime = [False] * N

for x in xrange(1, nsqrt):
    for y in xrange(1, nsqrt):
        n = 4 * (x * x) + (y * y)
        if n <= N and n % 12 in [1,5]:
            is_prime[n] = True

        n = 3 * (x * x) + (y * y)
        if n <= N and n % 12 is 7:
            is_prime[n] = True

        n = 3 * (x * x) - (y * y)
        if n <= N and x > y and n % 12 is 11:
            is_prime[n] = True

is_prime[2] = True
is_prime[3] = True

primes = []
for k, v in enumerate(is_prime):
    v and primes.append(k)

print(len(primes))
